### ISSUE TITLE
SyntaxError: `return try {`

### ISSUE THREAD

**Issue by vais** [opal/opal#1305]
The following Ruby:
```ruby
  it "can be used to handle a specific exception" do
    begin
      raise SpecificExampleException, "Raising this to be handled below"
    rescue SpecificExampleException
      :caught
    end.should == :caught
  end
```
compiles to the following JavaScript:
```javascript
($a = ($c = self).$it, $a.$$p = (TMP_3 = function(){var self = TMP_3.$$s || this;
    return try {
        return self.$raise($scope.get('SpecificExampleException'), "Raising this to be handled below")
      } catch ($err) {
        if (Opal.rescue($err, [$scope.get('SpecificExampleException')])) {
          try {
            return "caught"
          } finally { Opal.pop_exception() }
        } else { throw $err; }
      }.$should()['$==']("caught")}, TMP_3.$$s = self, TMP_3), $a).call($c, "can be used to handle a specific exception");
```
which, when executed, produces a SyntaxError:
```
    return try {
           ^^^

SyntaxError: Unexpected token try
    at exports.runInThisContext (vm.js:53:16)
    at Module._compile (module.js:374:25)
    at Object.Module._extensions..js (module.js:417:10)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
    at Function.Module.runMain (module.js:442:10)
    at startup (node.js:136:18)
    at node.js:966:3
```
Due to this issue, we can no longer execute https://github.com/ruby/spec/blob/master/language/rescue_spec.rb#L17

**Comment by iliabylich** [comment]
I guess for now we can just put `nil` to the end of every test like this (and that's probably what Opal compiler should do).

**Comment by iliabylich** [comment]
Ahh, this spec comes from the rubyspec :smile:

**Comment by vais** [comment]
@iliabylich lol, yes :)

**Comment by elia** [comment]
it fails when treated as an expression, starting a 0.10 milestone to keep track of this stuff
in the meanwhile I think it's better to pin down a working mspec version in a submodule and restore the green badge

**Comment by vais** [comment]
@elia
> restore the green badge

??

https://github.com/opal/opal/pull/1306 is green, but `language/rescue` spec had to be commented out because of this syntax error. Not sure why you're referring to MSpec though. I guess @eregon made a change so some of the evaled specs are easier to filter out. We could have done that without updating to latest spec suite, it involves just changing some existing filters to remove the embedded `\n`s.

**Comment by elia** [comment]
My understanding is that if I run the travis build on master it would now fail, because part of the build process is to `bundle update` thus updating `mspec` to the latest commit, producing the errors we both saw. The change I'm referring to is this:

https://github.com/ruby/mspec/commit/7a19a5997aa313fe0faf7ff6f2edba1cecb44654

which "forbids" using `should_not raise_error` (which is good btw).

---

What I suggest is to add mspec as a submodule and use a version prior to that change, and then take our time to sort out the new `rescue` bug before upgrading `ruby/spec` and `ruby/mspec`.

**Comment by eregon** [comment]
Both mspec and ruby/spec changed recently.
The mspec change should cause at worst new filters (good since they were likely passing for the wrong reasons), but no exclusion. MSpec tries to be fully backward compatible except when it's bugged :smile: 

I think the issue here is caused by https://github.com/ruby/spec/commit/25f9b873eb3f636ad10e32ed735ef0b7fb007967#diff-e2abdf033ad4e925f2fa95de0a133cf2.
I'm happy to add any other kind of expectation in that example, so to not also test `begin/rescue/end` return value but just that the exception is caught (we could have `:caught.should == :caught` in the rescue).
I sort of became allergic to "should_not raise_error" because it's in the great majority of cases a sign of bad specs.

In summary, feel free to simplify these changes so that they compile and do not rely on more features than before. I'm also fine to temporarily wrap them in `not_supported_on :opal` if you prefer that.

**Comment by elia** [comment]
@eregon @vais my bad for commenting on this issue about the real problem we had, which is that the specs unexpectedly broke (after a bundle update) because of this 

https://github.com/ruby/mspec/commit/7a19a5997aa313fe0faf7ff6f2edba1cecb44654

The correct fix of course being #1306 since MSpec was out of sync with Ruby Spec. The other needed fix was moving MSpec from the Gemfile to a git submodule and update it together with the Ruby Spec. 

https://github.com/opal/opal/commit/de887531884a191cc9d6099b1feafdbf6ea4d79b

With that in place we can take our time to sort this stuff out without haste :)

**Comment by vais** [comment]
Thanks @elia and @eregon. I guess not we just have to fix the Opal compiler to get rid of this issue :wink:

**Comment by iliabylich** [comment]
@vais There are two issues:
``` ruby
begin
rescue
end.object_id
```
Compiles to:
``` js
return try {
  return nil
} catch ($err) {
  if (Opal.rescue($err, [$scope.get('StandardError')])) {
    try {
      return nil
    } finally { Opal.pop_exception() }
  } else { throw $err; }
}.$object_id()
```

1. `return try {`
2. `catch {}.$object_id()`

Both constructions are invalid. The first one is easy to fix by modifying compiled code to `try {} catch {}; return nil;` in `JSReturnNode` (are you ok with this solution? if so, I can send a PR). The second one looks more difficult to fix, @elia any ideas? :smile:

**Comment by elia** [comment]
@iliabylich in theory it should wrap the `try/catch` in a function, but some how this check is not triggered:

https://github.com/opal/opal/blob/master/lib/opal/nodes/rescue.rb#L89
